### ISSUE TITLE
IPv6 problem with “get.jenkins.io”

### ISSUE THREAD

**Issue by DonEstefan** [jenkins-infra/helpdesk#3639]
### Service(s)

jenkins.io

### Summary

We noticed that the domain “[get.jenkins.io](http://get.jenkins.io/)” is involved in Jenkins update delivery and has an DNS AAAA-Record pointing to “2603:1030:408:7::44”. This IPv6-Address does not respond to https requests and does not deliver any updates. Since dual-stack clients are required by RfC to prefer IPv6 over IPv4, some Jenkins instances are effectively cut off from security update downloads.
 
“[get.jenkins.io](http://get.jenkins.io/)” does have an DNS A-Record for IPv4, which is working fine. But fallback from IPv6 to IPv4 can be so slow (up to 60 seconds), that some clients give up before the fallback mechanism kicks in. Some clients/proxies implement “happy eyeballs” mechanism (https://en.wikipedia.org/wiki/Happy_Eyeballs), to work around problems of non-responding IPv6 endpoints in dual-stack environments. Also `-Djava.net.preferIPv4Stack=true ` is a possible workaround for installations that don't use a http proxy. So, the effect is somewhat limited. But dual-stack clients/proxies without “happy eyeballs” currently have problem downloading patches.

For a proper fix I’d like to kindly ask you to remove the non-responding IPv6 Endpoints from DNS.

### Reproduction steps

```
curl -6 https://get.jenkins.io/ -v
* About to connect() to get.jenkins.io port 443 (#0)
*   Trying 2603:1030:408:7::44... Connection timed out
* couldn't connect to host
* Closing connection #0
curl: (7) couldn't connect to host
```

**Comment by dduportal** [comment]
Hi @DonEstefan , thanks for contacting us (and raising this issue)!
You are correct and a proper fix is needed.

We are working on this, most probably tomorrow trying to fix this.


----

Initial technical analysis:

- In https://github.com/jenkins-infra/helpdesk/issues/3351 we migrated services including `get.jenkins.io` to a new cluster with IPv6 support (Dualstack).
- [The public IPv6 associated](https://github.com/jenkins-infra/azure/blob/c57a68bc0a3045ae168c884b4ed90c676418db65/publick8s.tf#L164-L172) with the [`AAAA` record `get.jenkins.io`](https://github.com/jenkins-infra/azure/blob/c57a68bc0a3045ae168c884b4ed90c676418db65/publick8s.tf#L183-L190) is created and allocated
- This IPv6 is properly specified to the Public Loadbalancer setup as per https://github.com/jenkins-infra/kubernetes-management/blob/4ab07477fba367b2b77541c14ffa9b65e1529a11/config/public-nginx-ingress_publick8s.yaml#L9

```
$ dig AAAA get.jenkins.io

; <<>> DiG 9.10.6 <<>> AAAA get.jenkins.io
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 3022
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 4, ADDITIONAL: 9

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 1232
;; QUESTION SECTION:
;get.jenkins.io.                        IN      AAAA

;; ANSWER SECTION:
get.jenkins.io.         60      IN      CNAME   public.publick8s.jenkins.io.
public.publick8s.jenkins.io. 300 IN     AAAA    2603:1030:408:7::44
```

=> We are clearly missing some part of the network setup and need to learn and dig deeper

**Comment by dduportal** [comment]
Hi @DonEstefan , with https://github.com/jenkins-infra/azure/pull/417, we'll delete the current `AAAA` record for `get.jenkins.io`  as a temporary solution to unblock your organization.

We'll add it back once we'll have fully tested and persisted the IPv6 setup.

**Comment by dduportal** [comment]
I'm reopening as we just validated a working setup for IPv6: we should apply it to `get.jenkins.io` later today.

@DonEstefan Expect some hiccups if we miss a step, but the setup we just validated in production was working well: you should be able to reach `get.jenkins.io` through IPv6 later today (the IP should be the same).

**Comment by dduportal** [comment]
Technical analysis (following up and end):

https://learn.microsoft.com/en-us/azure/aks/configure-kubenet-dual-stack?tabs=azure-cli%2Ckubectl#expose-the-workload-via-a-loadbalancer-type-service tells us:

> 2. Only the first IP address for a service will be provisioned to the load balancer, so a dual-stack service only receives a public IP for its first-listed IP family. To provide a dual-stack service for a single deployment, please create two services targeting the same selector, one for IPv4 and one for IPv6.

- As the current (and only) Service LoadBalacner created by the ingress-nginx charts specifies dualstask with IPv4 as the first item in the list, it created internal and external IPv4 addresses but no IPv6 => it explains why the service was not reachable
- We verified the following elements manually on the cluster:
  - Nginx deployment responds properly to Ipv6 ✅ 
  - We can create an additional LoadBalancer, forced to DualStack network but with only Ipv6, which allocates both external and internal Ipv6 ✅ 
  - The external Ipv6, once associated to an `AAAA` DNS records is properly reachable by the outside world ✅ 
  - We can specify the actual IPv6 (defined in Terraform) in the new Service LB annotations: it reuses it without error instead of creating a new one (we deleted and re-created the IPv6 LB to verify this) ✅

**Comment by dduportal** [comment]
Update:
- New helm chart introduced in https://github.com/jenkins-infra/helm-charts/pull/544 to create the IPv6 LoadBalancer
- Let's delay to tomorrow (29 June) for the next step to prepare a proper configuration as code setup (no blocker in opening the PR in kubernetes-management though @lemeurherve, it's the merge that we should do in pair/mob)

**Comment by lemeurherve** [comment]
A LB service dedicated to IPv6 has been deployed to `publick8s` cluster hosting get.jenkins.io service, and the AAAA record has been restored.

Checked via https://ready.chair6.net/?url=https%3A%2F%2Fget.jenkins.io, the issue should be properly fixed now.

![image](https://github.com/jenkins-infra/helpdesk/assets/91831478/05ce5b7f-0b72-4542-b102-551614d094d3)

@DonEstefan could you make another round of tests please?

**Comment by dduportal** [comment]
Thanks @DonEstefan for the feedback and for the contribution by providing all the precious informations you reported! 

Happy that it works 🥳

**Comment by lemeurherveCB** [comment]
> Oh I got it: as per https://learn.microsoft.com/en-us/azure/aks/configure-kubenet-dual-stack?tabs=azure-cli%2Ckubectl#expose-the-workload-via-a-loadbalancer-type-service (in the admonition):
> 
> > Only the first IP address for a service will be provisioned to the load balancer, so a dual-stack service only receives a public IP for its first-listed IP family. To provide a dual-stack service for a single deployment, please create two services targeting the same selector, one for IPv4 and one for IPv6.
> 
> That means that the setup for our ingress, assuming the Service of type LB, only serves IPv4. (Ping @lemeurherve It also explains the weird behavior you met during the very first deployment of the ingress that forced: I guess the IPv6 was provisionned first).

@dduportal FTR and according to https://learn.microsoft.com/en-gb/azure/aks/azure-cni-overlay?tabs=kubectl#expose-the-workload-via-a-loadbalancer-type-service there is no need for a dedicated IPv6 LB anymore with Kubernetes 1.27 and later: Prior to Kubernetes version 1.27, only the first IP address for a service will be provisioned to the load balancer, so a dual-stack service only receives a public IP for its first-listed IP family. To provide a dual-stack service for a single deployment, please create two services targeting the same selector, one for IPv4 and one for IPv6. This is no longer a limitation in kubernetes 1.27 or later.

Current deployment to be revised accordingly? (Nice-to-have, low priority)

**Comment by dduportal** [comment]
> > Oh I got it: as per https://learn.microsoft.com/en-us/azure/aks/configure-kubenet-dual-stack?tabs=azure-cli%2Ckubectl#expose-the-workload-via-a-loadbalancer-type-service (in the admonition):
> > > Only the first IP address for a service will be provisioned to the load balancer, so a dual-stack service only receives a public IP for its first-listed IP family. To provide a dual-stack service for a single deployment, please create two services targeting the same selector, one for IPv4 and one for IPv6.
> > 
> > 
> > That means that the setup for our ingress, assuming the Service of type LB, only serves IPv4. (Ping @lemeurherve It also explains the weird behavior you met during the very first deployment of the ingress that forced: I guess the IPv6 was provisionned first).
> 
> @dduportal FTR and according to https://learn.microsoft.com/en-gb/azure/aks/azure-cni-overlay?tabs=kubectl#expose-the-workload-via-a-loadbalancer-type-service there is no need for a dedicated IPv6 LB anymore with Kubernetes 1.27 and later:
> 
> > Prior to Kubernetes version 1.27, only the first IP address for a service will be provisioned to the load balancer, so a dual-stack service only receives a public IP for its first-listed IP family. To provide a dual-stack service for a single deployment, please create two services targeting the same selector, one for IPv4 and one for IPv6. This is no longer a limitation in kubernetes 1.27 or later.
> 
> Current deployment to be revised accordingly? (Nice-to-have, low priority)

Worth testing it on a brand new Ingress installation:

- The "Service LB" problem we had seems resolved as there are annotation for the public IP: https://cloud-provider-azure.sigs.k8s.io/topics/loadbalancer/#loadbalancer-annotations (as your links are only describing dynamically allocated IP which is different than our production)
- I don't know if Nginx Ingress Helm chart will set up the proper labels though
